### ISSUE TITLE
New Feature - Tenant-Level Browsing

### ISSUE THREAD

**Issue by jimmywim** [tomdam/spcoder#2]
This is more an FYI rather than issue. I've got a requirement to run a version of the existing Workflow getter script against a whole tenant. So I've added a TenantNode object and extended the CSOMConnector to expend the tenant and get all child sites. So far going well (and I've created a modified Workflow Getter to iterate Tenant Sites of the target object).

Only annoyance at the minute is that the "Tenant" appears twice in the hierarchy. Apart from that, everything works fine. I think it's just down to the logic in the ExplorerView that renders the rootNode.

![image](https://user-images.githubusercontent.com/3036301/88061441-64b2ea00-cb5f-11ea-980c-1156aeeb571b.png)

**Comment by jimmywim** [comment]
Just an update, I got this fixed. I have merged in other changes into the branch I'm working in that fixes both List & Library expansion, and shows the list of Content Types.

![image](https://user-images.githubusercontent.com/3036301/88067068-6e8c1b80-cb66-11ea-98fa-8d44d1196944.png)

Having the Tenant node available as a target also opens up the possibility of your scripts doing stuff at the Tenant level, too (like my example of Workflow getter working against all site collections). 

See PR: https://github.com/tomdam/spcoder/pull/3

~~(You'll need to demonstrate how to get `ExplorerView` to be able to access `SharePointClientModule` without needing to explicitly add it in code, which as you've said creates an unwanted dependency on that project DLL).~~ *Nevermind, figure it out by doing a full Solution build which includes the SharePoint.Client assembly properly into the output dir*

**Comment by tomdam** [comment]
Great work! This is a really useful feature to have.
I'm testing it and will merge your PR afterwards.

**Comment by tomdam** [comment]
PR is merged now. I've noticed that it wasn't possible to expand the subwebs after connecting to tenant, but I found a fix and commited it.

**Comment by jimmywim** [comment]
Ah sweet, I found bugs on expanding folders once I implemented the tenant stuff and merged the content type stuff, forgot to test subwebs!

**Comment by tomdam** [comment]
That's fine. It was a great contribution from your side!
My next idea for CSOM connector is to create container nodes similar to ContentType container node that you have created, to show fields, flows, workflows (2010 & 2013), views, event receivers... for lists/libraries. Practically any "important" object that could then be added to context and viewed with describer/properties viewer/gridviewer...
After that, the same could be done with site/web properties.

**Comment by jimmywim** [comment]
One thing I've noticed about the context system - is there a way currently we can hydrate an object more fully when added to context? Currently, I've populated objects in the TreeView using the bare needed to get the Tree View to work (quickly, reducing ExecuteQueries when looping items when parent nodes are expanded), but that leaves it sparse when inspected in the properties viewer. Maybe BaseNode could do with a "EnsureAllProperties" method that's called on an object when added to context?

**Comment by tomdam** [comment]
My idea for that was to add a new menu item for the context items (below rename) that would get all properties from server. So, a user would drop an item to context and, if needed, call that action to ensure properties.
It can also be done after an object has been added to context, but we should test what would happen in situations when ExecuteQuery takes long time to finish... 

Adding the "EnsureAllProperties" method to BaseNode should be the right way to do it... the rest would just be the decision on when to call it.

**Comment by jimmywim** [comment]
True, that sounds a sensible way to do it. Maybe should have a dialog or some progress indicator when it is taking a while.

A final Q for now on this app - is there an easy way to write to the Log window from the other assembles? I noticed log4net is used (SPCoderLogging.Logger?) but that only writes to file. Just thinking of what the best pattern for writing traces in things like the Connector classes.

**Comment by tomdam** [comment]
Good question... currently there's no direct method that could be called to write to log window from other assemblies. 
The way to implement that would be through a callback that could be registered in SPCoder.Core (since it is used by all the other projects). I'll put that task to my todo list.

**Comment by tomdam** [comment]
Hi, just wanted to let you know that I added a way to write to log window from other Modules(Connectors). You can call methods of SPCoderLogger.Logger object to do that. It implements ISPCoderLog interface so you can add error, info and warning entries.
